rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+    ) -> Result<LockItem, DynamoError> {
+        let lookup_time = now_millis();
+        let rvn = Uuid::new_v4().to_string();
+
+        let mut item = hashmap! {
+            PARTITION_KEY_NAME.to_string() => attr(self.opts.partition_key_value.clone()),
+            OWNER_NAME.to_string() => attr(&self.opts.owner_name),
+            RECORD_VERSION_NUMBER.to_string() => attr(&rvn),
+            LEASE_DURATION.to_string() => attr(&self.opts.lease_duration),
+        };
+
+        if let Some(ref d) = data {
+            item.insert(DATA.to_string(), attr(d));
+        }
+
+        self.client
+            .put_item(PutItemInput {
+                table_name: self.opts.table_name.clone(),
+                item,
+                condition_expression,
+                expression_attribute_names,
+                expression_attribute_values,
+                ..Default::default()
+            })
+            .await?;
+
+        Ok(LockItem {
+            owner_name: self.opts.owner_name.clone(),
+            record_version_number: rvn,
+            lease_duration: self.opts.lease_duration,
+            is_released: false,
+            data,
+            lookup_time,
+        })
+    }
+}
+
+impl LockItem {
+    fn is_expired(&self) -> bool {
+        if self.is_released {
+            return true;
+        }
+        now_millis() - self.lookup_time > (self.lease_duration as u128) * 1000
+    }
+}
+
+fn now_millis() -> u128 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis()
+}
+
+pub fn attr<T: ToString>(s: T) -> AttributeValue {
+    AttributeValue {
+        s: Some(s.to_string()),
+        ..Default::default()
+    }
+}
+
+struct AcquireLockState<'a> {
+    client: &'a DynamoLock,
+    active_lock: Option<LockItem>,
+    started: Instant,
+    timeout_in: Duration,
+}
+
+impl<'a> AcquireLockState<'a> {
+    fn has_timed_out(&self) -> bool {
+        self.started.elapsed() > self.timeout_in
+    }
+
+    async fn try_acquire_lock(&mut self) -> Result<LockItem, DynamoError> {
+        match self.client.get_lock().await? {
+            None => {
+                // there's no lock, we good to acquire it
+                Ok(self.upsert_new_lock().await?)
+            }
+            Some(existing) if existing.is_released => {
+                // lock is released by a caller, we good to acquire it
+                Ok(self.upsert_released_lock(existing.data).await?)
+            }
+            Some(existing) => {
+                // there's existing lock and it's out first attempt to acquire it
+                if self.active_lock.is_none() {
+                    // first we store it, extend timeout period and try again later
+                    return self.set_new_active_lock_and_fail(existing);
+                }
+
+                // there's existing lock and we've already tried to acquire it, let's try again
+                let active = self.active_lock.as_ref().unwrap();
```

Comment:
You can probably get rid of the unwrap here by using a match assign pattern:

```suggestion
                let active = match self.active_lock.as_ref() {
                    Some(active) => active,
                    None => {
                        // first we store it, extend timeout period and try again later
                        return self.set_new_active_lock_and_fail(existing);
                    }
                };
```